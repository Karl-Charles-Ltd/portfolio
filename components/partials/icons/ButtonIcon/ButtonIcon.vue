<template>
  <div :class="outerClasses" class="v-btn-icon">
    <component :is="tag" class="v-btn-icon__btn" tabindex="0" v-bind="$attrs" v-on="buttonListeners">
      <BaseIcon
        class="v-btn-icon__graphic"
        :direction="direction"
        :icon="graphic"
        :size="size"
        :theme="theme"
      />
    </component>
  </div>
</template>

<script lang="ts">
import { Prop, Component, mixins } from 'nuxt-property-decorator';
import iconMixin from '../iconMixin.vue';
import * as icons from '../icons';
import buttonMixin from '~/components/partials/buttons/buttonMixin.vue';

const availableSizes: string[] = ['teeny', 'tiny', 'small', 'base', 'medium', 'intermediate', 'big'];

@Component({
  inheritAttrs: false,
  components: {},
})
export default class ButtonIcon extends mixins(iconMixin, buttonMixin) {
  @Prop({ required: true, validator: (value) => Object.keys(icons).includes(value) }) readonly icon!: string;
  @Prop({ default: 'small', validator: (value) => availableSizes.includes(value) })
  readonly size!: string;

  get graphic() {
    return this.loading && this.disc ? 'LoadingSpinner' : this.icon;
  }

  get outerClasses() {
    return {
      [`v-btn-icon--${this.size}`]: this.size,
      [`v-btn-icon--${this.theme}`]: this.theme,
      'v-btn-icon--focus': this.focus,
      'v-btn-icon--disabled': this.disabled,
      'v-btn-icon--disc': this.disc,
      'v-btn-icon--busy': this.loading,
    };
  }
}
</script>

<style lang="scss">
@import '../setup.scss';

.v-btn-icon {
  border-radius: 50%;
  display: inline-block;
  position: relative;
  transition: box-shadow 0.2s ease-in;
  vertical-align: top;

  &--disabled {
    .v-btn-icon__btn {
      cursor: unset;
      opacity: 0.5;
      pointer-events: none;
    }
  }

  &--busy {
    .v-btn-icon__btn {
      cursor: wait;
      opacity: 0.5;
      pointer-events: none;
    }
  }

  &__btn {
    background-color: transparent;
    border: none;
    cursor: pointer;
    left: 50%;
    padding: 0;
    position: absolute;
    top: 50%;
    transform: translate(-50%, -50%);
    z-index: z-index('foreground');
  }

  &__graphic {
    transition: background-color 0.2s ease-in;
  }
}

@each $_size-name, $_size-value in $disc-icon-size-map {
  .v-btn-icon--#{$_size-name} {
    height: calc(#{$_size-value} + 4px);
    width: calc(#{$_size-value} + 4px);
  }
}

@each $_theme-name, $_theme-map in $button-icon-style-map {
  .v-btn-icon--#{$_theme-name}.v-btn-icon--disc {
    .v-btn-icon__btn {
      outline: none;
    }

    @each $_state, $_color in $_theme-map {
      &.v-btn-icon--focus {
        box-shadow: 0 0 2px 2px map-get($_theme-map, 'focus');
      }
      .v-btn-icon__btn:hover .v-btn-icon__graphic:not(.v-btn-icon__btn--disabled) {
        background-color: map-get($_theme-map, 'hover') !important;
      }
    }
  }
}
</style>
