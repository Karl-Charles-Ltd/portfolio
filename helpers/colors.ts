/**
 * Get the hex color codes that are used throughout the project.
 * These colors and names should match with
 * assets/scss/settings/_settings.colors.scss.
 *
 * @param {String} colorName - The color name from the theme
 * @returns {String} - The hex code for the matching color
 */
export function colorFromTheme(colorName) {
  const colors = {
    primary: '#370b3c',
    secondary: '#6b2f93',
    tertiary: '#D0438F',
    quaternary: '#f40ee3',
    quinary: '#09ECFD',
    'quinary-light': '#1A5C7E',
    'quinary-dark': '#010226',
    black: '#000000',
    background: '#202227',
    charcoal: '#4D4D4D',
    'grey-one': '#949494',
    'grey-two': '#CDCDCD',
    'light-grey': '#E2E2E2',
    'light-grey-one': '#E2E2E2',
    'light-grey-two': '#F9F9F9',
    white: '#FFFFFF',
  };

  return colors[colorName];
}

export const themeMap = {
  primary: {
    background: 'primary',
    foreground: 'secondary',
  },
  'primary-dark': {
    background: 'primary-dark',
    foreground: 'secondary',
  },
  secondary: {
    background: 'secondary',
    foreground: 'white',
    accent: 'tertiary',
  },
  'secondary-dark': {
    background: 'secondary-dark',
    foreground: 'white',
  },
  'secondary-light': {
    background: 'secondary-light',
    foreground: 'white',
  },
  tertiary: {
    background: 'tertiary',
    foreground: 'secondary',
  },
  quaternary: {
    background: 'quaternary',
    foreground: 'secondary',
  },
  quinary: {
    background: 'quinary',
    foreground: 'white',
  },
  'quinary-darker': {
    background: 'quinary-darker',
    foreground: 'white',
  },
  white: {
    background: 'white',
    foreground: 'secondary',
  },
  charcoal: {
    background: 'charcoal',
    foreground: 'white',
  },
  'light-grey-one': {
    background: 'light-grey-one',
    foreground: 'secondary',
  },
};
