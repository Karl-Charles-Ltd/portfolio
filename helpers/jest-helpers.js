import { shallowMount, createLocalVue } from '@vue/test-utils';
import { createRenderer } from 'vue-server-renderer';
import VueRouter from 'vue-router';

export function setup() {
  const renderer = createRenderer();
  const localVue = createLocalVue();
  const router = new VueRouter();

  // setupPlugins(useVuex, localVue);

  return {
    renderer,
    router,
    localVue,
    shallowMount,
  };
}

export function componentFactory(component, options) {
  return shallowMount(component, options);
}

export function createSnapshot(component, options) {
  const wrapper = componentFactory(component, options);
  expect(wrapper.exists()).toBeTruthy();
  expect(wrapper.html()).toMatchSnapshot();
}
