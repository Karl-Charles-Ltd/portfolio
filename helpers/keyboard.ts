export function addKeyupEventListener(keyCode, callBack) {
  if (!window) return;
  window.addEventListener('keyup', (event) => {
    if (event.keyCode === keyCode) callBack();
  });
}

export function removeKeyupEventListener(keyCode, callBack) {
  if (!window) return;
  window.removeEventListener('keyup', (event) => {
    if (event.keyCode === keyCode) callBack();
  });
}
