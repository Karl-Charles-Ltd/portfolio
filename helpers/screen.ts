import * as screenBreakpoints from '~/constants/screen-breakpoints';

export const breakpoints: string[] = [
  screenBreakpoints.MOBILE,
  screenBreakpoints.TABLET,
  screenBreakpoints.SMALL_DESK,
  screenBreakpoints.LARGE_DESK,
  screenBreakpoints.HUGE_DESK,
];

interface Screen {
  [key: string]: string | null;
}

export default function mq({ from = null, until = null }: Screen, currentScreenSize: string): boolean {
  const currentBreakpointIndex = breakpoints.indexOf(currentScreenSize);

  if (from !== '' && from !== null) {
    const fromBreakpointIndex = breakpoints.indexOf(from || '');
    if (fromBreakpointIndex > currentBreakpointIndex) {
      return false;
    }
  }

  if (until !== '' && until !== null) {
    const untilBreakpointIndex = breakpoints.indexOf(until || '');
    if (untilBreakpointIndex <= currentBreakpointIndex) {
      return false;
    }
  }

  return true;
}
