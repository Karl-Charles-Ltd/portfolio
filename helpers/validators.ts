import { themeMap } from './colors';

export function validateTheme(theme: string, extraColors: string[] = []): boolean {
  const baseThemes: string[] = Object.keys(themeMap);
  const availableThemes: string[] = baseThemes.concat(extraColors);

  return availableThemes.includes(theme);
}
