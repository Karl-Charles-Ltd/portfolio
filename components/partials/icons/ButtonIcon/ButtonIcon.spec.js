import ButtonIcon from './ButtonIcon.vue';
import { createSnapshot, testComputedProperty } from '~/helpers/jest-helpers';

describe('ButtonIcon', () => {
  test('renders a button with a base icon', () => {
    createSnapshot(ButtonIcon, { propsData: { icon: 'Tick', disc: false }, stubs: ['BaseIcon'] });
  });

  describe('computed', () => {
    describe('outerClasses property', () => {
      let tempVue;
      const expectedOuterClasses = {
        size: 'v-btn-icon--intermediate',
        focus: 'v-btn-icon--focus',
        disc: 'v-btn-icon--disc',
        disabled: 'v-btn-icon--disabled',
        theme: 'v-btn-icon--quinary',
      };
      beforeEach(() => {
        tempVue = {
          size: 'intermediate',
          theme: 'quinary',
          focus: true,
          disabled: true,
          disc: true,
        };
      });
      it('returns correct icon size', () => {
        const outerClasses = testComputedProperty(ButtonIcon, 'outerClasses', tempVue);
        expect(outerClasses[expectedOuterClasses.size]).toBeTruthy();
      });
      it('returns correct icon theme', () => {
        const outerClasses = testComputedProperty(ButtonIcon, 'outerClasses', tempVue);
        expect(outerClasses[expectedOuterClasses.theme]).toBeTruthy();
      });
      it('returns focus class', () => {
        const outerClasses = testComputedProperty(ButtonIcon, 'outerClasses', tempVue);
        expect(outerClasses[expectedOuterClasses.focus]).toBeTruthy();
      });
      it('returns disabled class', () => {
        const outerClasses = testComputedProperty(ButtonIcon, 'outerClasses', tempVue);
        expect(outerClasses[expectedOuterClasses.disabled]).toBeTruthy();
      });
      it('returns disc class', () => {
        const outerClasses = testComputedProperty(ButtonIcon, 'outerClasses', tempVue);
        expect(outerClasses[expectedOuterClasses.disc]).toBeTruthy();
      });
    });
    describe('graphic property', () => {
      let tempVue;
      beforeEach(() => {
        tempVue = {
          icon: 'Test',
          loading: true,
          disc: true,
        };
      });
      it('returns loading spinner if loading and disc are true icon', () => {
        const graphic = testComputedProperty(ButtonIcon, 'graphic', tempVue);
        expect(graphic).toEqual('LoadingSpinner');
      });
      it('returns icon if loading or disc are false', () => {
        tempVue.loading = false;
        const graphic = testComputedProperty(ButtonIcon, 'graphic', tempVue);
        expect(graphic).toEqual('Test');
      });
    });
  });
});
