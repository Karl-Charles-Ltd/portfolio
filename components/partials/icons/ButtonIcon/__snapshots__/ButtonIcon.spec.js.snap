// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ButtonIcon renders a button with a base icon 1`] = `
"<div class=\\"v-btn-icon v-btn-icon--small v-btn-icon--secondary\\"><button tabindex=\\"0\\" class=\\"v-btn-icon__btn\\">
    <baseicon-stub icon=\\"Tick\\" size=\\"small\\" theme=\\"secondary\\" class=\\"v-btn-icon__graphic\\"></baseicon-stub>
  </button></div>"
`;
